pkg/idp/provider/github/github.go

```diff
@@ -0,0 +1,252 @@
+	return g.Type
+}
+
+func (g *Github) GetOwner() string {
+	return g.Owner
+}
+
+func (g *Github) CreateOrUpdateApp(config provider.AppConfig, ctx context.Context, oldConnector dex.Connector) (provider.ProviderApp, error) {
+	secret, err := g.createOrUpdateSecret(config, ctx, oldConnector)
+	if err != nil {
+		return provider.ProviderApp{}, microerror.Mask(err)
+	}
+
+	connectorConfig := &githubconnector.Config{
+		ClientID:     secret.ClientId,
+		ClientSecret: secret.ClientSecret,
+		Orgs: []githubconnector.Org{
+			{
+				Name:  g.Organization,
+				Teams: []string{g.Team},
+			},
+		},
+		RedirectURI: config.RedirectURI,
+	}
+	data, err := yaml.Marshal(connectorConfig)
+	if err != nil {
+		return provider.ProviderApp{}, microerror.Mask(err)
+	}
+	return provider.ProviderApp{
+		Connector: dex.Connector{
+			Type:   g.Type,
+			ID:     g.Name,
+			Name:   key.GetConnectorDescription(ProviderConnectorType, g.Owner),
+			Config: string(data[:]),
+		},
+		SecretEndDateTime: secret.EndDateTime,
+	}, nil
+}
+
+func (g *Github) DeleteApp(name string, ctx context.Context) error {
+	// get authenticated app
+	_, _, err := g.Client.Apps.Get(ctx, "")
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	//TODO: remove redirect URI for the affected app
+	return nil
+}
+
+func (g *Github) createOrUpdateSecret(config provider.AppConfig, ctx context.Context, oldConnector dex.Connector) (provider.ProviderSecret, error) {
+	// get authenticated app, check if the callback URI is present
+	app, _, err := g.Client.Apps.Get(ctx, "")
+	if err != nil {
+		return provider.ProviderSecret{}, microerror.Mask(err)
+	}
+	if !callbackURIPresent(app, config) {
+		//We return here since we can not automatically set the URI
+		return provider.ProviderSecret{}, microerror.Maskf(missingCallbackURIError, fmt.Sprintf("Callback URI %s is not registered in %s app %s for %s in organization %s.", config.RedirectURI, g.Type, app.GetSlug(), g.Owner, g.Organization))
+	}
+	if g.UpdateNeeded(app, config) {
+		//We return here since we can not set the update
+		return provider.ProviderSecret{}, microerror.Maskf(missingCallbackURIError, fmt.Sprintf("%s app %s for %s in github organization %s needs update.", g.Type, app.GetSlug(), g.Owner, g.Organization))
+	}
+	return g.getSecret(app, oldConnector)
+}
+
+func (g *Github) getSecret(app *githubclient.App, oldConnector dex.Connector) (provider.ProviderSecret, error) {
+	var err error
+	var endDateTime time.Time
+	var clientID, clientSecret string
+	{
+		// Retrieve old id and secret
+		clientID, clientSecret, err = getSecretFromConfig(oldConnector.Config)
+		if err != nil {
+			return provider.ProviderSecret{}, microerror.Mask(err)
+		}
+		//currently the authenticated app represents the MC dex app
+		//WC case can be handled by adding extra callback URIs to the authenticated app
+		if clientID != g.id || clientSecret != g.secret {
+			clientID = g.id
+			clientSecret = g.secret
+		}
+		endDateTime = app.GetCreatedAt().AddDate(0, 6, 0)
```

Comment:
We consider the app expired 6 month after creation which will force us to renew it manually (since rotation is not supported by the API)
However, a CLI for easier manual renewal will be provided